Remove some unsafeGet/unsafePtr usage from WebCore/dom

annevk opened the pull request:
#### 3025232f837965ab567f9de4330bb82b29866cfd
<pre>
Remove some unsafeGet/unsafePtr usage from WebCore/dom
<a href="https://bugs.webkit.org/show_bug.cgi?id=301414">https://bugs.webkit.org/show_bug.cgi?id=301414</a>

Reviewed by Ryosuke Niwa.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines</a> (as
they are expected to evolve).

Canonical link: <a href="https://commits.webkit.org/302162@main">https://commits.webkit.org/302162@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9fb00dd3de6a73222797322fecfa1ed8afd13927

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128128 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/412 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38960 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135493 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79623 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a6142b13-d45f-4775-8337-47e5b5680f3f) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130000 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/333 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/284 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97537 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65432 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/76e2cd14-e0ee-4bd2-be89-25639611130b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131076 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/201 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114777 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78107 "Found 1 new API test failure: WPE/TestWebsiteData:/webkit/WebKitWebsiteData/itp (failure)") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9db6fac4-6099-4578-8143-f520c3b865d4) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/191 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32884 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78804 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108567 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33370 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137983 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/264 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/248 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106064 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/295 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111120 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105803 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/199 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29674 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52483 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20032 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/311 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/62111 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/220 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/293 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/270 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -10911,9 +10911,9 @@ void Document::removeTopLayerElement(Element& element)
-        if (RefPtr dialog = dynamicDowncast<HTMLDialogElement>(element.get()); dialog && dialog->isModal())
-            return dialog.unsafeGet();
+    for (Ref element : makeReversedRange(m_topLayerElements)) {
+        if (auto* dialog = dynamicDowncast<HTMLDialogElement>(element.get()); dialog && dialog->isModal())
```
I don't think this is safe. If element was the last `Ref` to this element, then we'd return a pointer value to freed memory. Why does element need to be `Ref` at all? It seems like everything being called here is trivial so the whole thing should be considered as trivial.

annevk replied:
`m_topLayerElements` is a vector of `Ref` so it won't be anything but a `Ref`, right?

rniwa replied:
We should be able to get it as `Ref&` as we were.

annevk replied:
I see, I didn't realize that was distinct from `Ref`!

rniwa commented on Source/WebCore/dom/DocumentFragment.cpp:
```diff
@@ -114,19 +114,19 @@ bool DocumentFragment::parseXML(const String& source, Element* contextElement, O
     for (Ref element : descendantsOfType<Element>(*this)) {
         if (element->getIdAttribute() == id)
-            return const_cast<Element*>(element.unsafePtr());
+            return const_cast<Element*>(element.ptr());
```
We should return `element` instead. That'll avoid a ref churn.

annevk replied:
I moved the `const_cast` to `this`. I tried it here with `Element&` but that didn't work.

annevk committed:
Remove some unsafeGet/unsafePtr usage from WebCore/dom
https://bugs.webkit.org/show_bug.cgi?id=301414

Reviewed by Ryosuke Niwa.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines (as
they are expected to evolve).

Canonical link: https://commits.webkit.org/302162@main
